### PR TITLE
jit: add constraint for torch.device as input to jitted fn

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: #555 

This PR adds constraints to the prologue trace to check the passed torch.device to jitted function matches the previously traced torch.device.

**NOTE**: We add this constraint even for `SYMBOLIC_VALUES` caching option, as the computation trace may have operations which are supported only on certain devices, so we shouldn't be reusing that trace.

Example:
```python
import thunder
import torch

def foo(x, device):
    return x.to(device)

jfoo = thunder.jit(foo)
x = torch.randn(3, 3)
print(jfoo(x, torch.device('cuda')).device)  # output : cuda:0
print("****")
print(thunder.last_prologue_traces(jfoo)[-1])
```

Prologue Trace (after the PR):
```python
# Constructed by Transform for execution (took 1 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def prologue(*args, **kwargs):
  # args: "Any"
  check_len(args, 2)
    # prims.check_len(args, 2)
  # kwargs: "Any"
  check_len(kwargs, 0)
    # prims.check_len(kwargs, 0)
  x: "cpu f32[3, 3]" = args[0]
  check_tensor_metadata(x, (3, 3), 'cpu', torch.float32, False)
    # prims.check_tensor_metadata(x, (3, 3), 'cpu', torch.float32, False)

  # THIS CHECK IS ADDED BY THE PR
  p0: "<class 'torch.device'>" = args[1]
  check_literal_like(p0, torch.device("cuda"))
    # prims.check_literal_like(p0, torch.device("cuda"))

  cache_info: "Any" = thunder._get_cache_info()
  cache_info_is_autocast_enabled: "bool False" = cache_info['is_autocast_enabled']
  check_number_type_and_value(cache_info_is_autocast_enabled, False)
    # prims.check_number_type_and_value(cache_info_is_autocast_enabled, False)
  cache_info_no_grad_sync: "bool False" = cache_info['no_grad_sync']
  check_number_type_and_value(cache_info_no_grad_sync, False)
    # prims.check_number_type_and_value(cache_info_no_grad_sync, False)
  return (x,)
```

### CODE REVIEW

File: thunder/core/jit_ext.py [hunk 1]
```diff
@@ -895,7 +907,7 @@ def proxy_recursion(v):
             v.track_items()
             need_proxy = any(proxy_recursion(i) for i in v.item_wrappers)
         else:
-            need_proxy = isinstance(v.value, torch.Tensor)
+            need_proxy = isinstance(v.value, torch.Tensor) or isinstance(v.value, torch.device)
```
Comment (t-vi) [inline thunder/core/jit_ext.py:910]: I might be wrong, but could we just add the check but not proxy just like for `int` / `str`?

Comment (kshitij12345) [reply thunder/core/jit_ext.py]: It works (locally verified on the test). I have applied it in https://github.com/Lightning-AI/lightning-thunder/pull/556/commits/2073cb41384bcdae00c0baeaddbfbd95dd251ea7. Is this what you had in mind? Thanks!

File: thunder/core/jit_ext.py [hunk 2]
```diff
@@ -560,6 +561,17 @@ def proxify(self, value: WrappedValue) -> Any:
         # avoid double registration by skipping if value has a registered proxy.
         if isinstance(uvalue, Proxy) or value.original_value is not value.nothing:
             return uvalue
+        elif isinstance(uvalue, torch.device):
+            co: CACHE_OPTIONS = get_cache_option()
+            p: AnyProxy = proxy(uvalue, history=value.provenance)
+            if co in (CACHE_OPTIONS.CONSTANT_VALUES, CACHE_OPTIONS.SYMBOLIC_VALUES):
+                # NOTE: Even with SYMBOLIC_VALUES, we want to strictly constraint the device as
+                # the computation trace may utilize device specific executors.
+                self.add_constraint((clang.check_literal_like, p, uvalue))
```
Comment (kshitij12345) [inline thunder/core/jit_ext.py:570]: cc: @jjsjann123 for opinion and sanity check.

Comment (jjsjann123) [reply thunder/core/jit_ext.py:570]: :clap:

**Review by jjsjann123 — APPROVED** [review]
LGTM

**Pull request merged**